File: Leetcode/cpp/sol560.cpp

//https://leetcode.com/problems/subarray-sum-equals-k/description/

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> prefix_map = {{0,1}};
        int prefix_sum = 0;
        int count = 0;
        for(int num:nums){
            prefix_sum+=num; //add the current sum of the array
            int needed = prefix_sum - k; // find a previous sum in the array so substracting would get us k
            if(prefix_map.count(needed)){
                count+=prefix_map[needed];
            }
            prefix_map[prefix_sum]+=1;
        }

        return count;

    }
};

File: Leetcode/python/sol560.py
# https://leetcode.com/problems/subarray-sum-equals-k/description/

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        if len(nums) < 1:
            return 0

        prefix_sum = 0
        prefix_map = {0:1}
        count = 0

        for num in nums:
            prefix_sum+=num
            needed =  prefix_sum - k
            if needed in prefix_map:
                count += prefix_map[needed]
            sums[prefix_sum] = sums.get(prefix_sum, 0)+1
        return count

